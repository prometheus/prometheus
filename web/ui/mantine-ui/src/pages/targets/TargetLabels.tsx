import { FC } from "react";
import { Labels } from "../../api/responseTypes/targets";
import { LabelBadges } from "../../components/LabelBadges";
import { ActionIcon, Collapse, Group, Stack, Text } from "@mantine/core";
import { useDisclosure } from "@mantine/hooks";
import { IconChevronDown, IconChevronUp } from "@tabler/icons-react";

type TargetLabelsProps = {
  labels: Labels;
  discoveredLabels: Labels;
};

const TargetLabels: FC<TargetLabelsProps> = ({ discoveredLabels, labels }) => {
  const [showDiscovered, { toggle: toggleDiscovered }] = useDisclosure(false);

  return (
    <Stack>
      <Group wrap="nowrap" align="flex-start">
        <LabelBadges labels={labels} />

        <ActionIcon
          size="xs"
          color="gray"
          variant="light"
          onClick={toggleDiscovered}
          title={`${showDiscovered ? "Hide" : "Show"} discovered (pre-relabeling) labels`}
        >
          {showDiscovered ? (
            <IconChevronUp
              style={{ width: "70%", height: "70%" }}
              stroke={1.5}
            />
          ) : (
            <IconChevronDown style={{ width: "70%", height: "70%" }} />
          )}
        </ActionIcon>
      </Group>

      <Collapse in={showDiscovered}>
        <Text fw={700} size="1em" my="lg" c="gray.7">
          Discovered labels:
        </Text>
        <LabelBadges color="blue" labels={discoveredLabels} />
      </Collapse>
    </Stack>
  );
};

export default TargetLabels;
