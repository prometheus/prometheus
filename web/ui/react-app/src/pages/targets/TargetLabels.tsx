import React, { FC, Fragment, useState } from 'react';
import { Badge, Tooltip } from 'reactstrap';
import 'css.escape';
import styles from './TargetLabels.module.css';

interface Labels {
  [key: string]: string;
}

export interface TargetLabelsProps {
  discoveredLabels: Labels;
  labels: Labels;
  idx: number;
  scrapePool: string;
}

const formatLabels = (labels: Labels): string[] => Object.keys(labels).map(key => `${key}="${labels[key]}"`);

const TargetLabels: FC<TargetLabelsProps> = ({ discoveredLabels, labels, idx, scrapePool }) => {
  const [tooltipOpen, setTooltipOpen] = useState(false);

  const toggle = (): void => setTooltipOpen(!tooltipOpen);
  const id = `series-labels-${scrapePool}-${idx}`;

  return (
    <>
      <div id={id} className="series-labels-container">
        {Object.keys(labels).map(labelName => {
          return (
            <Badge color="primary" className="mr-1" key={labelName}>
              {`${labelName}="${labels[labelName]}"`}
            </Badge>
          );
        })}
      </div>
      <Tooltip isOpen={tooltipOpen} target={CSS.escape(id)} toggle={toggle} style={{ maxWidth: 'none', textAlign: 'left' }}>
        <b>Before relabeling:</b>
        {formatLabels(discoveredLabels).map((s: string, idx: number) => (
          <Fragment key={idx}>
            <br />
            <span className={styles.discovered}>{s}</span>
          </Fragment>
        ))}
      </Tooltip>
    </>
  );
};

export default TargetLabels;
