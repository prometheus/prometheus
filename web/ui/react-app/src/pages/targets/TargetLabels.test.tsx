import * as React from 'react';
import { shallow } from 'enzyme';
import TargetLabels from './TargetLabels';
import { Tooltip, Badge } from 'reactstrap';
import toJson from 'enzyme-to-json';

describe('targetLabels', () => {
  const defaultProps = {
    discoveredLabels: {
      __address__: 'localhost:9100',
      __metrics_path__: '/metrics',
      __scheme__: 'http',
      job: 'node_exporter',
    },
    labels: {
      instance: 'localhost:9100',
      job: 'node_exporter',
      foo: 'bar',
    },
    idx: 1,
    scrapePool: 'cortex/node-exporter_group/0',
  };
  const targetLabels = shallow(<TargetLabels {...defaultProps} />);

  it('renders a div of series labels', () => {
    const div = targetLabels.find('div').filterWhere(elem => elem.hasClass('series-labels-container'));
    expect(div).toHaveLength(1);
    expect(div.prop('id')).toEqual('series-labels-cortex/node-exporter_group/0-1');
  });

  it('wraps each label in a label badge', () => {
    const l: { [key: string]: string } = defaultProps.labels;
    Object.keys(l).forEach((labelName: string): void => {
      const badge = targetLabels
        .find(Badge)
        .filterWhere(badge => badge.children().text() === `${labelName}="${l[labelName]}"`);
      expect(badge).toHaveLength(1);
    });
    expect(targetLabels.find(Badge)).toHaveLength(3);
  });

  it('renders a tooltip for discovered labels', () => {
    const tooltip = targetLabels.find(Tooltip);
    expect(tooltip).toHaveLength(1);
    expect(tooltip.prop('isOpen')).toBe(false);
    expect(tooltip.prop('target')).toEqual('series-labels-cortex\\/node-exporter_group\\/0-1');
  });

  it('renders discovered labels', () => {
    expect(toJson(targetLabels)).toMatchSnapshot();
  });
});
