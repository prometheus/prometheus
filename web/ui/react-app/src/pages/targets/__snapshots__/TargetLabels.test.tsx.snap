// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`targetLabels renders discovered labels 1`] = `
<Fragment>
  <div
    className="series-labels-container"
    id="series-labels-cortex/node-exporter_group/0-1"
  >
    <Badge
      className="mr-1"
      color="primary"
      key="instance"
      pill={false}
      tag="span"
    >
      instance="localhost:9100"
    </Badge>
    <Badge
      className="mr-1"
      color="primary"
      key="job"
      pill={false}
      tag="span"
    >
      job="node_exporter"
    </Badge>
    <Badge
      className="mr-1"
      color="primary"
      key="foo"
      pill={false}
      tag="span"
    >
      foo="bar"
    </Badge>
  </div>
  <Tooltip
    autohide={true}
    isOpen={false}
    placement="top"
    placementPrefix="bs-tooltip"
    style={
      Object {
        "maxWidth": "none",
        "textAlign": "left",
      }
    }
    target="series-labels-cortex\\\\/node-exporter_group\\\\/0-1"
    toggle={[Function]}
    trigger="hover focus"
  >
    <b>
      Before relabeling:
    </b>
    <br />
    <span
      className="discovered"
    >
      __address__="localhost:9100"
    </span>
    <br />
    <span
      className="discovered"
    >
      __metrics_path__="/metrics"
    </span>
    <br />
    <span
      className="discovered"
    >
      __scheme__="http"
    </span>
    <br />
    <span
      className="discovered"
    >
      job="node_exporter"
    </span>
  </Tooltip>
</Fragment>
`;
